CrossSectionAnalysis/cpp/CrossSectionCompute.cxx

```diff
@@ -0,0 +1,340 @@
+            crossSectionProperties->SetInputData(polydata);
+            const double crossSectionSurfaceArea = crossSectionProperties->GetSurfaceArea();
+            const double ceDiameter = (sqrt(crossSectionSurfaceArea / vtkMath::Pi())) * 2;
+            
+            bufferArray->InsertNextTuple3((double) i, crossSectionSurfaceArea, ceDiameter);
+        }
+        polydata->Delete();
+    }
+    
+    // Manage last point for an input centerline model
+    if (inputCenterlineNode->IsA("vtkMRMLModelNode"))
+    {
+        // Cross-section area is not calculated at last point for a model.
+        vtkMRMLModelNode * inputModel = vtkMRMLModelNode::SafeDownCast(inputCenterlineNode);
+        vtkPoints * modelPoints = inputModel->GetMesh()->GetPoints();
+        const unsigned int numberOfPoints = modelPoints->GetNumberOfPoints();
+        if (endPointIndex == (numberOfPoints - 1))
+        {
+            const unsigned int blockLastPointIndex = bufferArray->GetNumberOfTuples() - 1;
+            const unsigned int blockBeforeLastPointIndex = bufferArray->GetNumberOfTuples() - 1 - 1;
+            bufferArray->SetTuple3(blockLastPointIndex,
+                                   bufferArray->GetTuple3(blockLastPointIndex)[0],
+                                   bufferArray->GetTuple3(blockBeforeLastPointIndex)[1],
+                                   bufferArray->GetTuple3(blockBeforeLastPointIndex)[2]);
+        }
+    }
+    
+    // The centerline copy is not needed anymore.
+    inputCenterlineNode->Delete();
+    
+    #if DEVTIME != 0
+    mtx.lock();
+    double endTime = GetTimeOfDay();;
+    cout << "This thread : " << std::this_thread::get_id() << " "
+        << (endTime - startTime) / 1000000 << " seconds" << endl;
+    mtx.unlock();
+    #endif
+}
+
+// Translated and adapted from the Python implementation.
+vtkPolyData * vtkCrossSectionComputeWorker::ComputeCrossSectionPolydata(
+                                vtkMRMLNode * inputCenterlineNode,
+                                vtkPolyData * closedSurfacePolyData,
+                                unsigned int pointIndex)
+{
+    if (inputCenterlineNode == NULL)
+    {
+        WORKER_MESSAGE("Input centerline is NULL.");
+        return NULL;
+    }
+    if (closedSurfacePolyData == NULL)
+    {
+        WORKER_MESSAGE("Closed  surface polydata is NULL.");
+        return NULL;
+    }
+    
+    double center[3] = {0.0, 0.0, 0.0};
+    double normal[3] = {0.0, 0.0, 0.0};
+    if (inputCenterlineNode->IsA("vtkMRMLModelNode"))
+    {
+        vtkMRMLModelNode * inputModel = vtkMRMLModelNode::SafeDownCast(inputCenterlineNode);
+        vtkPoints * modelPoints = inputModel->GetMesh()->GetPoints();
+        const unsigned int numberOfPoints = modelPoints->GetNumberOfPoints();
+        double centerLocal[3] = {0.0, 0.0, 0.0};
+        modelPoints->GetPoint(pointIndex, centerLocal);
+        inputModel->TransformPointToWorld(centerLocal, center);
```

Comment:
Transform the model and input surface to world coordinate system once, outside the threads. In the threads just process the resulting VTK objects.